### DISCUSSION TITLE
Brother HL-4150CDN series (initial IPP info)

### DISCUSSION THREAD

**Discussion by plvlnisse** [OpenPrinting/libcups#93] [Show and tell]
So I got some time to look into my Brother printer and how to make it survive into CUPS 3 and onwards. I suspect that many other Brother printers may share capabilities with this one, I thought I'd share what I can get from it as of now while trying to learn more about IPP and CUPS.

So with the below information is the recommended path forward to implement support either by using an "IPP proxy" such as PAPPL or pappl-retrofit or is there a way to piggy-back on availability of other Brother printers in the CUPS-3 PPD database and how those printers are handled (or not handled) in CUPS3 and beyond?

The printer says it supports IPP 1.0 in its response, but IPP-tool is unhappy that responses are i HTML-1.0. It understands if I send the question below using IPP-1.1, but croaks on IPP-2.0. 

The printer itself supports input in (at least) PostScript, HP-PCL and Brother-Script v3 as I understand it.

```
cat ask-printer.txt 
{
 OPERATION Get-Printer-Attributes

 GROUP operation-attributes-tag
 ATTR charset "attributes-charset" "utf-8"
 ATTR naturalLanguage "attributes-natural-language" "en"
 ATTR uri "printer-uri" "$uri"
 ATTR name "requesting-user-name" "John Doe"
 # ATTR mimeMediaType "document-format" "application/pdf"
 ATTR keyword "requested-attributes" "printer-description","job-template","media-col-database"
}
```
``ipptool -h -S -C -l -V 1.0 -v -t https://192.168.0.111/ ask-printer.txt`` results in:

```
"ask-printer.txt":
    Get-Printer-Attributes:
        attributes-charset (charset) = utf-8
        attributes-natural-language (naturalLanguage) = en
        printer-uri (uri) = https://192.168.0.111:443/
        requesting-user-name (nameWithoutLanguage) = John Doe
        requested-attributes (1setOf keyword) = printer-description,job-template,media-col-database
    ask-printer                                                          [FAIL]
        RECEIVED: 1847 bytes in response
        status-code = successful-ok-ignored-or-substituted-attributes (successful-ok-ignored-or-substituted-attributes)
        attributes-charset (charset) = utf-8
        attributes-natural-language (naturalLanguage) = en-us
        requested-attributes (keyword) = media-col-database
        generated-natural-language-supported (1setOf naturalLanguage) = en-us,da,nl,fi,fr,de,it,no,pt,sv,es,tr,pl,ja,ru,cs,hu
        printer-uri-supported (1setOf uri) = http://192.168.0.111:631/ipp,http://192.168.0.111:631/ipp/port1,https://192.168.0.111/ipp,https://192.168.0.111/ipp/port1,https://192.168.0.111/ipp,https://192.168.0.111/ipp/port1,http://brother:631/ipp,http://brother:631/ipp/port1,https://brother/ipp,https://brother/ipp/port1,https://brother/ipp,https://brother/ipp/port1
        uri-security-supported (1setOf keyword) = none,none,ssl3,ssl3,tls,tls,none,none,ssl3,ssl3,tls,tls
        uri-authentication-supported (1setOf keyword) = none,none,none,none,none,none,none,none,none,none,none,none
        printer-name (nameWithoutLanguage) = brother
        printer-location (no-value) = no-value
        printer-info (textWithoutLanguage) = MFG:Brother;CMD:PJL,PCL,PCLXL;MDL:HL-4150CDN series;CLS:PRINTER;CID:Brother Color Type3;
        printer-make-and-model (textWithoutLanguage) = Brother HL-4150CDN series
        printer-state (enum) = idle
        printer-state-reasons (keyword) = none
        printer-state-message (textWithoutLanguage) = Sleep
        ipp-version-supported (keyword) = 1.0
        operations-supported (1setOf enum) = Print-Job,Validate-Job,Create-Job,Send-Document,Cancel-Job,Get-Job-Attributes,Get-Jobs,Get-Printer-Attributes
        multiple-document-jobs-supported (boolean) = true
        multiple-operation-time-out (integer) = 150
        natural-language-configured (naturalLanguage) = en-us
        charset-configured (charset) = utf-8
        charset-supported (charset) = utf-8
        document-format-supported (1setOf mimeMediaType) = application/vnd.hp-PCL,application/postscript,application/vnd.brother-hbp,application/octet-stream
        document-format-default (mimeMediaType) = application/octet-stream
        printer-is-accepting-jobs (boolean) = true
        queued-job-count (integer) = 0
        pdl-override-supported (keyword) = not-attempted
        printer-up-time (integer) = 185617
        compression-supported (keyword) = none
        Bad HTTP version (1.0)

```

**Comment by michaelrsweet** [comment]
So this Brother printer is definitely using one of the first-generation IPP implementations that basically were an alternative to LPD or AppSocket.

Long-term you really have just two options:

- [hp-printer-app](https://github.com/michaelrsweet/hp-printer-app) which will send PCL to the printer, or
- [ps-printer-app](https://github.com/OpenPrinting/ps-printer-app) which will send PostScript to the printer.

"hp-printer-app" will probably provide the fastest printing overall.

**Comment by plvlnisse** [comment]
I can understand that this is a working option, however there are already a whole lot of Brother printers supported in the Gutenprint driver. So I would ideally (at least) start to just add the printer into this collection in the hope that the Gutenprint project will have their supported printers carried over into driverless in a larger porting effort. 

By just adding a line for my printer into /usr/share/gutenprint/5.3/xml/printers/pcl.xml using a similar Brother printer as a template I managed to get my printer recognized in the CUPS GUI and setup using the standard Gutenprint option in the GUI (no separate driver and PPD downloaded from Brother and installed as a separate filter). Some things are not correct, such as the number of trays, a scaling issue and the selectable Default Options - but it works!

Looking through the CUPS and Gutenprint sources I can not find where the printer information comes from - in particular the ".drv" database. I have not been able to find out where the "ppds.dat" database comes from and if that is where the particulares of an individual printer should be added - or if it is part of the "StpDriverModelFamily" 50001_pcl/62_pcl. But I guess I should start to bug the Gutenprint community first and join effort with them to carry all/most the Brother PostScript printers (and a bunch of other old PCL/PJL support) over to driverless CUPS.

**Reply by michaelrsweet** [reply]
Gutenprint is a completely separate project from this one... There is an existing "wrapper" printer application as well as some future work towards a more "native" solution.

WRT where the printer information comes from, there is a driver interface program that is supplied with Gutenprint which lists the available drivers and generates the corresponding PPD files.
